x-pack/plugins/security/server/authentication/authenticator.ts

```diff
@@ -600,4 +659,46 @@ export class Authenticator {
 
     return { idleTimeoutExpiration, lifespanExpiration };
   }
+
+  /**
+   * Checks whether request should be redirected to the Login Selector UI.
+   * @param request Request instance.
+   * @param session Current session value if any.
+   */
+  private shouldRedirectToLoginSelector(request: KibanaRequest, session: ProviderSession | null) {
+    // Request should be redirected to Login Selector UI only if all following conditions are met:
+    //  1. Request can be redirected (not API call)
+    //  2. Request is not authenticated yet
+    //  3. Login Selector UI is enabled
+    //  4. Request isn't attributed with HTTP Authorization header
+    return (
+      canRedirectRequest(request) &&
+      !session &&
+      this.options.config.authc.selector.enabled &&
+      HTTPAuthorizationHeader.parseFromRequest(request) == null
+    );
+  }
+
+  /**
+   * Checks whether request should be redirected to the Access Notice UI.
+   * @param request Request instance.
+   * @param session Current session value if any.
+   */
+  private shouldRedirectToAccessNotice(request: KibanaRequest, session: ProviderSession | null) {
+    // Request should be redirected to Access Notice UI only if all following conditions are met:
+    //  1. Request can be redirected (not API call)
+    //  2. Request is authenticated, but user hasn't acknowledged access notice in the current
+    //     session yet (based on the flag we store in the session)
+    //  3. Request is authenticated by the provider that has `accessNotice` configured
+    //  4. And it's not a request to the Access Notice UI itself
+    return (
+      canRedirectRequest(request) &&
+      session != null &&
+      !session.flags?.accessNoticeAcknowledged &&
+      (this.options.config.authc.providers as Record<string, any>)[session.provider.type]?.[
+        session.provider.name
+      ]?.accessNotice &&
+      request.url.pathname !== ACCESS_NOTICE_ROUTE
```

Comment:
__note:__ technically `pathname` can have a trailing slash, but I'm ignoring this here to simplify the check because of the following reasons:

* When we redirect to this page we don't add trailing slash
* If user goes directly to this page and adds a trailing slash, they will end up at `/security/access_notice?next=%2Fsecurity%2Faccess_notice%2F` assuming they haven't acknowledged access notice yet, that means after acknowledgement they will be redirect to `/security/access_notice` (since Kibana server will eventually issue a redirected to the same URL but without trailing slash), after second acknowledgement or just by navigating to another page everything will go back to normal.

I'm not worried about the second point since it's a user mistake after all and it doesn't have that bad consequences.